pytorch/2. Autograd.ipynb

```diff
@@ -0,0 +1,988 @@
+      "[torch.FloatTensor of size 3x10]\n",
+      " \n",
+      "do/db : \n",
+      "\n",
+      " Variable containing:\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      " 5\n",
+      "[torch.FloatTensor of size 10]\n",
+      " \n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Gradients of o w.r.t inputs in Graph 1')\n",
+    "print('do/dx : \\n\\n %s ' % (do_dinputs_1[0]))\n",
+    "print('do/dw : \\n\\n %s ' % (do_dinputs_1[1]))\n",
+    "print('do/db : \\n\\n %s ' % (do_dinputs_1[2]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Graph 2 : wx / b"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "o = torch.matmul(x, w) / b\n",
+    "do_dinputs_2 = torch.autograd.grad(o, [x, w, b], grad_outputs=torch.ones(5, 10))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Gradients of o w.r.t inputs in Graph 2\n",
+      "do/dx : \n",
+      " Variable containing:\n",
+      " 47.6666 -31.6417 -54.9581\n",
+      " 47.6666 -31.6417 -54.9581\n",
+      " 47.6666 -31.6417 -54.9581\n",
+      " 47.6666 -31.6417 -54.9581\n",
+      " 47.6666 -31.6417 -54.9581\n",
+      "[torch.FloatTensor of size 5x3]\n",
+      " \n",
+      "do/dw : \n",
+      " Variable containing:\n",
+      "\n",
+      "Columns 0 to 7 \n",
+      " 25.7204  -1.4251   0.5816   0.7336   0.3829  -0.5467   0.3904   0.3968\n",
+      "-85.3804   4.7306  -1.9306  -2.4353  -1.2709   1.8149  -1.2960  -1.3172\n",
+      " 97.1318  -5.3817   2.1963   2.7705   1.4459  -2.0647   1.4743   1.4985\n",
+      "\n",
+      "Columns 8 to 9 \n",
+      " -4.2812  -0.4352\n",
+      " 14.2118   1.4446\n",
+      "-16.1679  -1.6434\n",
+      "[torch.FloatTensor of size 3x10]\n",
+      " \n",
+      "do/db : \n",
+      " Variable containing:\n",
+      "-173.4977\n",
+      "   5.7707\n",
+      "   5.5410\n",
+      "  -2.9863\n",
+      "  -0.3088\n",
+      "  -1.7527\n",
+      "  -0.5234\n",
+      "  -1.5729\n",
+      "-261.3063\n",
+      "  -0.8931\n",
+      "[torch.FloatTensor of size 10]\n",
+      " \n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Gradients of o w.r.t inputs in Graph 2')\n",
+    "print('do/dx : \\n %s ' % (do_dinputs_2[0]))\n",
+    "print('do/dw : \\n %s ' % (do_dinputs_2[1]))\n",
+    "print('do/db : \\n %s ' % (do_dinputs_2[2]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Gradient buffers: `.backward()` and `retain_graph=True`\n",
+    "\n",
+    "1. Calling `.backward()` clears gradient buffers and \"erases the tape\".\n",
```

Comment:
I find it a bit confusing, since you say later that "Calling .backward() multiple times will accumulate gradients into .grad and NOT overwrite them". Is the "gradient buffer" something different from ".grad"?

Reply:
Also, maybe introduce `zero_grad()` here?

Reply:
Ah yes what I really meant is clears the tape. I shouldn't use the word gradient buffer. Presenting `zero_grad()` will probably require me talking about `torch.optim` since I believe the only way to clear the `.grad` attribute of a variable that isn't associated with an optimizer is ` variable.grad.data.zero_()`